### PR TITLE
Replaced mfxVideoParams with smart wrappers in SMT

### PR DESCRIPTION
**Pull request by dmitryermilov**
Signed-off-by: Dmitry Ermilov <dmitry.ermilov@intel.com>

### CODE REVIEW

File: samples/sample_multi_transcode/include/pipeline_transcode.h [hunk 1]
```diff
@@ -664,9 +662,18 @@ namespace TranscodingSample
         virtual mfxStatus PreEncPreInit(sInputParams *pParams);
 
         mfxVideoParam GetDecodeParam();
-        mfxExtOpaqueSurfaceAlloc GetDecOpaqueAlloc() const { return m_pmfxVPP.get() ? m_VppOpaqueAlloc: m_DecOpaqueAlloc; };
 
-        mfxExtMVCSeqDesc GetDecMVCSeqDesc() const {return m_MVCSeqDesc;}
+        mfxExtOpaqueSurfaceAlloc GetDecOpaqueAlloc()
+        {
+            mfxExtOpaqueSurfaceAlloc* opaq = m_pmfxVPP.get() ? m_mfxVppParams : m_mfxDecParams;
```
Comment (mgonchar) [inline samples/sample_multi_transcode/include/pipeline_transcode.h]: no need to call ```.get()``` to check pointer. all smart pointers have ```bool operator ()``` which makes their behavior really similar to regular pointers

File: samples/sample_multi_transcode/include/pipeline_transcode.h [hunk 2]
```diff
@@ -778,83 +785,35 @@ namespace TranscodingSample
...
-
-        // external parameters for each component are stored in a vector
-        CVPPExtBuffersStorage        m_VppExtParamsStorage;
-        std::vector<mfxExtBuffer*> m_EncExtParams;
-        std::vector<mfxExtBuffer*> m_DecExtParams;
-        std::vector<mfxExtBuffer*> m_PluginExtParams;
-        std::vector<mfxExtBuffer*> m_PreEncExtParams;
+        // to enable to-do list
+        // number of video enhancement filters (denoise, procamp, detail, video_analysis, multi_view, ste, istab, tcc, ace, svc)
+        #define ENH_FILTERS_COUNT                (20)
```
Comment (mgonchar) [inline samples/sample_multi_transcode/include/pipeline_transcode.h]: Better to use ```constexpr```

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp [hunk 3]
```diff
@@ -149,78 +149,12 @@ CTranscodingPipeline::CTranscodingPipeline():
...
-#endif
-
-#if (MFX_VERSION >= 1025)
-
-    MSDK_ZERO_MEMORY(m_ExtMFEParam);
-    MSDK_ZERO_MEMORY(m_ExtMFEControl);
-
-    m_ExtMFEControl.Header.BufferId = MFX_EXTBUFF_MULTI_FRAME_CONTROL;
-    m_ExtMFEControl.Header.BufferSz = sizeof(mfxExtMultiFrameControl);
+    m_RotateParam = {};
```
Comment (mgonchar) [inline samples/sample_multi_transcode/src/pipeline_transcode.cpp]: I'd move every trivial initialization in the constructor body to initialization list above

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp [hunk 4]
```diff
@@ -2206,40 +2110,39 @@ mfxStatus CTranscodingPipeline::AllocMVCSeqDesc()
...
-        MSDK_ZERO_MEMORY(m_MVCSeqDesc.View[i]);
+        mvc->View[i] = {};
     }
-    m_MVCSeqDesc.NumViewAlloc = m_MVCSeqDesc.NumView;
+    mvc->NumViewAlloc = mvc->NumView;
 
-    m_MVCSeqDesc.ViewId = new mfxU16[m_MVCSeqDesc.NumViewId];
-    MSDK_CHECK_POINTER(m_MVCSeqDesc.ViewId, MFX_ERR_MEMORY_ALLOC);
-    for (i = 0; i < m_MVCSeqDesc.NumViewId; ++i)
+    mvc->ViewId = new mfxU16[mvc->NumViewId];
```
Comment (mgonchar) [inline samples/sample_multi_transcode/src/pipeline_transcode.cpp]: place for next round of wrapping :)

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp [hunk 5]
```diff
@@ -2206,40 +2110,39 @@ mfxStatus CTranscodingPipeline::AllocMVCSeqDesc()
 {
     mfxU32 i;
 
-    m_MVCSeqDesc.View = new mfxMVCViewDependency[m_MVCSeqDesc.NumView];
-    MSDK_CHECK_POINTER(m_MVCSeqDesc.View, MFX_ERR_MEMORY_ALLOC);
-    for (i = 0; i < m_MVCSeqDesc.NumView; ++i)
+    auto mvc = m_mfxDecParams.AddExtBuffer<mfxExtMVCSeqDesc>();
+    mvc->View = new mfxMVCViewDependency[mvc->NumView];
```
Comment (mgonchar) [inline samples/sample_multi_transcode/src/pipeline_transcode.cpp]:
```
mvc->View = new mfxMVCViewDependency[mvc->NumView]();
```
should zero initialize allocated memory so you won't need to do it explicitly

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp [hunk 6]
```diff
@@ -2206,40 +2110,39 @@ mfxStatus CTranscodingPipeline::AllocMVCSeqDesc()
...
+    mvc->NumViewAlloc = mvc->NumView;
 
-    m_MVCSeqDesc.ViewId = new mfxU16[m_MVCSeqDesc.NumViewId];
-    MSDK_CHECK_POINTER(m_MVCSeqDesc.ViewId, MFX_ERR_MEMORY_ALLOC);
-    for (i = 0; i < m_MVCSeqDesc.NumViewId; ++i)
+    mvc->ViewId = new mfxU16[mvc->NumViewId];
+    for (i = 0; i < mvc->NumViewId; ++i)
     {
-        MSDK_ZERO_MEMORY(m_MVCSeqDesc.ViewId[i]);
+        mvc->ViewId[i] = {};
```
Comment (mgonchar) [inline samples/sample_multi_transcode/src/pipeline_transcode.cpp]: same about initialization

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp [hunk 7]
```diff
@@ -2206,40 +2110,39 @@ mfxStatus CTranscodingPipeline::AllocMVCSeqDesc()
...
+    mvc->NumViewIdAlloc = mvc->NumViewId;
 
-    m_MVCSeqDesc.OP = new mfxMVCOperationPoint[m_MVCSeqDesc.NumOP];
-    MSDK_CHECK_POINTER(m_MVCSeqDesc.OP, MFX_ERR_MEMORY_ALLOC);
-    for (i = 0; i < m_MVCSeqDesc.NumOP; ++i)
+    mvc->OP = new mfxMVCOperationPoint[mvc->NumOP];
+    for (i = 0; i < mvc->NumOP; ++i)
     {
-        MSDK_ZERO_MEMORY(m_MVCSeqDesc.OP[i]);
+        mvc->OP[i] = {};
```
Comment (mgonchar) [inline samples/sample_multi_transcode/src/pipeline_transcode.cpp]: same about initialization

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp [hunk 8]
```diff
@@ -2447,29 +2340,28 @@ mfxStatus CTranscodingPipeline::InitEncMfxParams(sInputParams *pInParams)
...
+        auto mfeCtrl = m_mfxEncParams.AddExtBuffer<mfxExtMultiFrameControl>();
+        mfeCtrl->Timeout = pInParams->mfeTimeout;
     }
 #endif
     if (m_pParentPipeline && m_pParentPipeline->m_pmfxPreENC.get())
     {
-        m_mfxEncParams.mfx.RateControlMethod       = MFX_RATECONTROL_LA_EXT;
-        m_mfxEncParams.mfx.EncodedOrder            = 1; // binary flag, 0 signals encoder to take frames in display order
-        m_mfxEncParams.AsyncDepth = m_mfxEncParams.AsyncDepth == 0 ? 2: m_mfxEncParams.AsyncDepth;
+        m_mfxEncParams.mfx.RateControlMethod = MFX_RATECONTROL_LA_EXT;
```
Comment (mgonchar) [inline samples/sample_multi_transcode/src/pipeline_transcode.cpp]: looks like you have occasionaly changed formatting in code below?

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp [hunk 9]
```diff
@@ -2609,47 +2501,44 @@ MFX_IOPATTERN_IN_VIDEO_MEMORY : MFX_IOPATTERN_IN_SYSTEM_MEMORY);
...
 
         if (pInParams->nExtBRC != EXTBRC_DEFAULT && (pInParams->EncodeId == MFX_CODEC_HEVC || pInParams->EncodeId == MFX_CODEC_AVC))
         {
-            m_CodingOption2.ExtBRC =(mfxU16)(pInParams->nExtBRC == EXTBRC_OFF ? MFX_CODINGOPTION_OFF : MFX_CODINGOPTION_ON);
+            co2->ExtBRC =(mfxU16)(pInParams->nExtBRC == EXTBRC_OFF ? MFX_CODINGOPTION_OFF : MFX_CODINGOPTION_ON);
         }
         else
         {
-            m_CodingOption2.ExtBRC = 0;
+            co2->ExtBRC = 0;
```
Comment (mgonchar) [inline samples/sample_multi_transcode/src/pipeline_transcode.cpp]: let's use named const instead. This one is ```MFX_CODINGOPTION_UNKNOWN```

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp [hunk 10]
```diff
@@ -2789,11 +2669,13 @@ mfxStatus CTranscodingPipeline::CorrectPreEncAuxPool(mfxU32 num_frames_in_pool)
 
 mfxStatus CTranscodingPipeline::AllocPreEncAuxPool()
 {
-    if (!m_pmfxPreENC.get()) return MFX_ERR_NONE;
+    if (!m_pmfxPreENC.get())
```
Comment (mgonchar) [inline samples/sample_multi_transcode/src/pipeline_transcode.cpp]: no need to call ```.get()```

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp [hunk 11]
```diff
@@ -4499,27 +4420,31 @@ mfxStatus CTranscodingPipeline::Reset()
 mfxStatus CTranscodingPipeline::AllocAndInitVppDoNotUse(sInputParams *pInParams)
 {
     std::vector<mfxU32> filtersDisabled;
-    if(pInParams->DenoiseLevel==-1)
+    if (pInParams->DenoiseLevel == -1)
     {
-        filtersDisabled.push_back(MFX_EXTBUFF_VPP_DENOISE); // turn off denoising (on by default)
+        filtersDisabled.emplace_back(MFX_EXTBUFF_VPP_DENOISE); // turn off denoising (on by default)
```
Comment (mgonchar) [inline samples/sample_multi_transcode/src/pipeline_transcode.cpp]: just note, no sense to switch to ```emplace_back``` here. For trivial types it brings no difference

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp [hunk 12]
```diff
@@ -4499,27 +4420,31 @@ mfxStatus CTranscodingPipeline::Reset()
...
+    {
+        auto doNotUse = m_mfxVppParams.AddExtBuffer<mfxExtVPPDoNotUse>();
+        doNotUse->NumAlg = (mfxU32)filtersDisabled.size();
 
-    if(m_VppDoNotUse.AlgList)
-        delete[] m_VppDoNotUse.AlgList;
-    m_VppDoNotUse.AlgList = new mfxU32 [m_VppDoNotUse.NumAlg];
-    MSDK_CHECK_POINTER(m_VppDoNotUse.AlgList,  MFX_ERR_MEMORY_ALLOC);
-    MSDK_MEMCPY(m_VppDoNotUse.AlgList,&filtersDisabled[0],sizeof(mfxU32)*filtersDisabled.size());
+        if (doNotUse->AlgList)
```
Comment (mgonchar) [inline samples/sample_multi_transcode/src/pipeline_transcode.cpp]: deleting ```nullptr``` is valid call, it just does nothing. So this check can be removed

File: samples/sample_multi_transcode/include/pipeline_transcode.h (hunk 1 above)
Comment (dmitryermilov) [reply samples/sample_multi_transcode/include/pipeline_transcode.h]: ok

File: samples/sample_multi_transcode/include/pipeline_transcode.h (hunk 2 above)
Comment (dmitryermilov) [reply samples/sample_multi_transcode/include/pipeline_transcode.h]: ok

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp (hunk 3 above)
Comment (dmitryermilov) [reply samples/sample_multi_transcode/src/pipeline_transcode.cpp]: Dmitry R. agains

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp (hunk 4 above)
Comment (dmitryermilov) [reply samples/sample_multi_transcode/src/pipeline_transcode.cpp]: Yes. But not in this commit

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp (hunk 5 above)
Comment (dmitryermilov) [reply samples/sample_multi_transcode/src/pipeline_transcode.cpp]: right. thanks.

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp (hunk 8 above)
Comment (dmitryermilov) [reply samples/sample_multi_transcode/src/pipeline_transcode.cpp]: on purpose

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp (hunk 9 above)
Comment (dmitryermilov) [reply samples/sample_multi_transcode/src/pipeline_transcode.cpp]: ok

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp (hunk 10 above)
Comment (dmitryermilov) [reply samples/sample_multi_transcode/src/pipeline_transcode.cpp]: ok

File: samples/sample_multi_transcode/src/pipeline_transcode.cpp (hunk 12 above)
Comment (dmitryermilov) [reply samples/sample_multi_transcode/src/pipeline_transcode.cpp]: ok

**Pull request merged**